### PR TITLE
refactor text renderer for better unit testability

### PR DESCRIPTION
**Pull request by christianparpart**

### CODE REVIEW

File: src/vtrasterizer/TextRenderer.cpp [hunk 1]
```diff
@@ -614,89 +554,75 @@ void TextRenderer::renderRasterizedGlyph(crispy::point pen,
...
+    auto const hash = hashTextAndStyle(codepoints, style);
+    text::shape_result const& glyphPositions =
+        getOrCreateCachedGlyphPositions(hash, codepoints, clusters, style);
+    crispy::point pen = _gridMetrics.mapBottomLeft(initialPenPosition);
+    auto const advanceX = *_gridMetrics.cellSize.width;
 
-            auto const hash = hashGlyphKeyAndPresentation(glyphPosition.glyph, glyphPosition.presentation);
+    for (text::glyph_position const& glyphPosition: glyphPositions)
+    {
+        if (AtlasTileAttributes const* attributes = ensureRasterizedIfDirectMapped(glyphPosition.glyph))
```
Comment (Yaraslaut) [inline src/vtrasterizer/TextRenderer.cpp:589]: we can replace it with 
```suggestion
        if (auto attributes = ensureRasterizedIfDirectMapped(glyphPosition.glyph))
```

**Review by Yaraslaut — COMMENTED** [review]
Few comments

File: src/vtbackend/primitives.h [hunk 2]
```diff
@@ -128,48 +128,15 @@ struct [[nodiscard]] CellLocation
...
-constexpr bool operator==(CellLocation a, CellLocation b) noexcept
-{
-    return a.line == b.line && a.column == b.column;
-}
-constexpr bool operator!=(CellLocation a, CellLocation b) noexcept
-{
-    return !(a == b);
-}
-
-constexpr bool operator<(CellLocation a, CellLocation b) noexcept
```
[the remark this answers is not in the document]
Comment (christianparpart) [reply src/vtbackend/primitives.h:147]: `<=>` implements `<`, `<=`, `==`, `!=`, `>=`, `>`. See https://en.cppreference.com/w/cpp/language/default_comparisons

This was IMHO the whole point for the spaceship operator, to reduce boilerplate.

[the remark this answers is not in the document]
Comment (Yaraslaut) [reply src/vtbackend/primitives.h:147]: I meant same logic for ` operator <`

File: src/vtrasterizer/TextRenderer.cpp [hunk 3]
```diff
@@ -614,89 +554,75 @@ void TextRenderer::renderRasterizedGlyph(crispy::point pen,
...
+void TextRenderer::renderTextGroup(std::u32string_view codepoints,
+                                   gsl::span<unsigned> clusters,
+                                   vtbackend::CellLocation initialPenPosition,
+                                   TextStyle style,
+                                   vtbackend::RGBColor color)
 {
-    if (!_textClusterGroup.codepoints.empty())
-    {
-        _textRendererEvents.onBeforeRenderingText();
+    if (codepoints.empty())
```
[the remark this answers is not in the document]
Comment (christianparpart) [reply src/vtrasterizer/TextRenderer.cpp:573]: `resetAndMovePenForward` has moved into the `TextGroupCluster`

File: src/vtbackend/primitives.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (christianparpart) [reply src/vtbackend/primitives.h:147]: Yes, see https://godbolt.org/z/aeEb8c3nG

[the remark this answers is not in the document]
Comment (Yaraslaut) [reply src/vtbackend/primitives.h:147]: Alright :)

**Pull request merged**